lib/src/components/AbstractDialog/AbstractDialog.tsx

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2019-present Sonatype, Inc.
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/.
+ */
+import React, {
+  forwardRef,
+  KeyboardEvent,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState
+} from 'react';
+
+import { omit } from 'ramda';
+import { Props, AbstractDialogContextValue } from './types';
+
+export const AbstractDialogContext = React.createContext<AbstractDialogContextValue | null>(null);
+
+// Typescript has rather obnoxiously partially removed support for HTMLDialogElement.
+// See https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1029#issuecomment-907490128 and
+// https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1029#issuecomment-968299542 . This even though
+// Firefox appears to intend to finally add support in v98. Until TS restores this, we'll have to jump
+// through some hoops...
+const hasWindow = typeof window !== 'undefined',
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    dynamicallyTypedWindow = hasWindow && window as any,
+    hasNativeModalSupport = !!(hasWindow && dynamicallyTypedWindow.HTMLDialogElement &&
+      dynamicallyTypedWindow.HTMLDialogElement.prototype.showModal);
+
+const createCancelEvent = () => new Event('cancel', { cancelable: true });
+
+/**
+ * Abstracted Dialog element implementation and behaviors.
+ * @param className - A classname string for the dialog element.
+ * @param onCancel - A callback function that gets called when the dialog is canceled.
+ * @param cancelOnClickOutside - A boolean that if true triggers onCancel when the user clicks outside the dialog.
+ * @param cancelOnClickOutsideTargetClassName - A classname, if specified, selects an element inside the dialog
+ *    element that becomes the reference element for when the user clicks outside of, triggers the cancelOnClickOutside.
+ * @param useNativeCancelOnEscape - If this is set to true, it will attempt to use native dialog element
+ *    cancel behavior when escape is pressed. By default this is set to false, which means it uses event.preventDefault
+ *    and calls onCancel callback when escape is pressed.
+ * @return - Abstracted dialog element.
+ */
+
+// propTypes static analysis doesn't work with the way this component is written
+/* eslint-disable react/prop-types */
+const AbstractDialog = forwardRef<HTMLDialogElement, Props>((props, ref) => {
+  const {
+    className,
+    children,
+    cancelOnClickOutside,
+    onCancel,
+    useNativeCancelOnEscape,
```

Comment:
Why does this need to be configurable? Why wouldn't they both have the same behavior for the ESC key?

Reply:
> Why does this need to be configurable? Why wouldn't they both have the same behavior for the ESC key?

Oh this is because we need to manage the animation when the user presses the ESC key for the Drawer component